chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala

```diff
@@ -136,7 +136,8 @@ object Vec extends VecFactory
 sealed class Vec[T <: Data] private[core] (gen: => T, val length: Int)
     extends Aggregate with VecLike[T] {
   override def toString: String = {
-    s"$sample_element[$length]$bindingToString"
+    val elementType = sample_element.cloneType
```

Comment:
Wait why is this necessary? This smells odd...

Reply:
Because [`bindingToString`](https://github.com/freechipsproject/chisel3/blob/c1ab9e7afd5072c11d879db913e1b553c7fe0dbe/chiselFrontend/src/main/scala/chisel3/core/Data.scala#L354) uses `topBindingOpt`. I changed `topBindingOpt` to return the actual top binding (which you want sometimes from a `sample_element`). So in order to keep the pretty printing sane: `UInt<8>[8](Wire in ...)` rather than `UInt<8>(Wire in ...)[8](Wire in ...)`, I thought it prudent to clone it to clear the binding.

Reply:
An alternative would be to make SampleElementBinding a TopBinding (it kind of is, at least in that it's not a hardware child of the container Vec), and have bindingToString for SampleElementBinding return something sane?

Reply:
I was thinking about that and I think it's a reasonable approach but there are a couple of issues that seem problematic (or at least weird) to me:

1. `SampleElementBinding` isn't truly a `TopBinding`, because it still needs to point to the `Vec` of which it's the sample element. One reason why this is true is discussed below,

2. We need to check that Analog is only bound to `Port`s and `Wire`s, not `Reg`s or `Mem`s. From the code:

```scala
    val targetTopBinding = target match {
      case target: TopBinding => target
      case ChildBinding(parent) => parent.topBinding
      // See https://github.com/freechipsproject/chisel3/pull/946
      case SampleElementBinding(parent) => parent.topBinding
    }

    // Analog counts as different directions based on binding context
    targetTopBinding match {
      case WireBinding(_) => direction = ActualDirection.Unspecified  // internal wire
      case PortBinding(_) => direction = ActualDirection.Bidirectional(ActualDirection.Default)
      case x => throwException(s"Analog can only be Ports and Wires, not '$x'")
    }
```
If we make `SampleElementBinding` a pseudo-`TopBinding`, we could rewrite it as something like:
```scala
    val targetTopBinding = target match {
      case SampleElementBinding(parent) => parent.topBinding // <-- Note this is also a TopBinding so has to be first
      case target: TopBinding => target
      case ChildBinding(parent) => parent.topBinding  
    }

    // Analog counts as different directions based on binding context
    targetTopBinding match {
      case WireBinding(_) => direction = ActualDirection.Unspecified  // internal wire
      case PortBinding(_) => direction = ActualDirection.Bidirectional(ActualDirection.Default)
      case x => throwException(s"Analog can only be Ports and Wires, not '$x'")
    }
```
We can't rely upon the non-sample elements for this check, because then the following (which should error) would not error when n == 0:
```scala
  val reg = Reg(Vec(n, Analog(1.W))
```

Reply:
I agree that the `cloneType` smells weird, but it smells less weird to me than the existence of `SampleElement` as only a pseudo-`TopBinding`

Reply:
Actually I think worse than what I said above because you could have nested Vecs as your type and thus you hit multiple `SampleElement` bindings on your way, so you'd have to do something like:

```scala
    val targetTopBinding = {
      def rec(binding: Binding): TopBinding = binding match {
        case SampleElementBinding(parent) => rec(parent.topBinding)
        case target: TopBinding => target
        case ChildBinding(parent) => rec(parent.topBinding)
      }
      rec(target)
    }
```
This makes it seem to me that `SampleElementBinding` doesn't make sense as a `TopBinding`.

Reply:
Yeah, this seems like a thorny question. I'm fine merging this hotfix as-is, and perhaps we should have a separate discussion on SampleElementBinding in its own issue.